File: Python/Problems/Problems/problem2593.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem2593(Problem):
    def test(self) -> bool:
        nums = [2, 1, 3, 4, 5, 2]

        expected = 7

        result = self.findScore(nums)

        print(result)

        return result == expected

    def findScore(self, nums: List[int]) -> int:
        n = len(nums)
        queue = []
        marked = [False] * (n + 2)
        for i in range(n):
            heapq.heappush(queue, [nums[i], i])

        result = 0
        while queue:
            item = heapq.heappop(queue)
            id = item[1]
            if not marked[id + 1]:
                marked[id] = True
                marked[id + 1] = True
                marked[id + 2] = True
                result += item[0]

        return result

File: C++/Problems/Problems/problem_2593.cpp
#include "problem_2593.h"

#include <queue>

bool problem_2593::test() {
    std::vector nums { 2, 1, 3, 4, 5, 2 };

    int expected = 7;

    int result = findScore(nums);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_2593::findScore(std::vector<int> &nums) {
    int n = nums.size();
    std::priority_queue<std::pair<int, int>> queue;
    std::vector marked(n + 2, false);
    for (int i = 0; i < n; i++)
        queue.push(std::pair(-nums[i], -i));

    long long result = 0;
    while (!queue.empty())
    {
        auto item = queue.top();
        queue.pop();
        auto id = -item.second;
        if (!marked[id+ 1])
        {
            marked[id] = true;
            marked[id + 1] = true;
            marked[id + 2] = true;
            result -= item.first;
        }
    }

    return result;
}
